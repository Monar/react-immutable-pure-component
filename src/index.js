/*
  Copyright (C) 2017 Piotr Tomasz Monarski.
  Licensed under the MIT License (MIT), see
  https://github.com/Monar/react-immutable-pure-component
*/

import React from 'react';
import { is } from 'immutable';


export class ImmutablePureComponent extends React.Component {

  shouldComponentUpdate(nextProps, nextState = {}) {
    const state = this.state || {};

    return !(this.updateOnProps || Object.keys({...nextProps, ...this.props})).every((p) => is(nextProps[p], this.props[p]))
      || !(this.updateOnStates || Object.keys({ ...nextState, ...state})).every((s) => is(nextState[s], state[s]));
  }
}

export default ImmutablePureComponent;
