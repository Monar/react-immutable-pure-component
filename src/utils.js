import { is, isImmutable, isOrdered } from 'immutable';

const NOT_SET = {};

export function check(config, prev = {}, next = {}, label) {
  let checkItem = createChecker(prev, next, label);
  let checklist = config || Object.keys({ ...next, ...prev });

  return checklist.every(checkItem);
}

function createChecker(prev, next, checkName) {
  return function(name) {
    if (typeof name === 'string') {
      return is(next[name], prev[name]);
    }

    if (!getIn) {
      let value = JSON.stringify(name);
      throw new TypeError(
        `Not supported value "${value}" provided to ${checkName}, try updating immutable to v4`,
      );
    }

    return is(getIn(next, name), getIn(prev, name));
  };
}

function get(collection, key, notSetValue) {
  return isImmutable(collection)
    ? collection.get(key, notSetValue)
    : typeof collection.get === 'function'
    ? collection.get(key, notSetValue)
    : hasOwnProperty.call(collection, key)
    ? collection[key]
    : notSetValue;
}

function coerceKeyPath(keyPath) {
  if (Array.isArray(keyPath) && typeof keyPath !== 'string') {
    return keyPath;
  }
  if (isOrdered(keyPath)) {
    return keyPath.toArray();
  }
  throw new TypeError(
    'Invalid keyPath: expected Ordered Collection or Array: ' + keyPath,
  );
}

function getIn(collection, searchKeyPath, notSetValue) {
  const keyPath = coerceKeyPath(searchKeyPath);
  let i = 0;
  while (i !== keyPath.length) {
    collection = get(collection, keyPath[i++], NOT_SET);
    if (collection === NOT_SET) {
      return notSetValue;
    }
  }
  return collection;
}
