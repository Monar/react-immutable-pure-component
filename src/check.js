import { is } from 'immutable';
import { getIn } from './utils';

export function check(config, prev = {}, next = {}, label) {
  let checkItem = createChecker(prev, next, label);
  let checklist = config || Object.keys({ ...next, ...prev });

  return checklist.every(checkItem);
}

function createChecker(prev, next) {
  return function(name) {
    if (typeof name === 'string') {
      return is(next[name], prev[name]);
    } else if (Array.isArray(name)) {
      return is(getIn(next, name), getIn(prev, name));
    }

    throw new TypeError('Invalid key: expected Array or string: ' + name);
  };
}
