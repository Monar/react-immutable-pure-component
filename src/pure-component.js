import React from 'react';
import { check } from './utils';

export class ImmutablePureComponent extends React.Component {
  shouldComponentUpdate(nextProps, nextState = {}) {
    return (
      !check(this.updateOnProps, this.props, nextProps, 'updateOnProps') ||
      !check(this.updateOnStates, this.state, nextState, 'updateOnStates')
    );
  }
}
