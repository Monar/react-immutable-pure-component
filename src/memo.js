import React from 'react';
import { check } from './check';

export function immutableMemo(Component, updateOnProps) {
  return React.memo(Component, (prev, next) =>
    check(updateOnProps, prev, next, 'immutableMemo'),
  );
}
