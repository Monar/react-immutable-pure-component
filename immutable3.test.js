/* eslint-env node, jest */
import { ImmutablePureComponent } from './src';

jest.mock('immutable', () => ({ is: () => true }));

const getTestComponent = (props = {}, state) => {
  const Component = new ImmutablePureComponent();
  Component.props = props;
  Component.state = state;
  return Component;
};

describe('check when immutable does not have getIn', () => {
  it('should throw when prop name is not a string', () => {
    const Component = getTestComponent({ a: 1 });
    Component.updateOnProps = [[]];
    expect(() =>
      Component.shouldComponentUpdate({ a: 1 }),
    ).toThrowErrorMatchingSnapshot();
  });

  it('should throw when state name is not a string', () => {
    const Component = getTestComponent({}, { a: 1 });
    Component.updateOnStates = [[]];
    expect(() =>
      Component.shouldComponentUpdate({}, { a: 1 }),
    ).toThrowErrorMatchingSnapshot();
  });
});
