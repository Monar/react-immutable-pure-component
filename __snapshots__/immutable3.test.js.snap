// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`check when immutable does not have getIn should throw when prop name is not a string 1`] = `"Not supported value \\"[]\\" provided to updateOnProps, try updating immutable to v4"`;

exports[`check when immutable does not have getIn should throw when state name is not a string 1`] = `"Not supported value \\"[]\\" provided to updateOnStates, try updating immutable to v4"`;
